api/commands/run.ts

```diff
@@ -61,134 +83,223 @@ export class Runner {
     if (!!this.executionTask) {
       throw new Error("Executor already started.");
     }
-
-    this.executionTask = new Promise(async (resolve, reject) => {
-      try {
-        // Work out all the schemas we are going to need to create first.
-        const uniqueSchemas: { [schema: string]: boolean } = {};
-        this.graph.actions
-          .filter(action => !!action.target)
-          .map(action => action.target.schema)
-          .filter(schema => !!schema)
-          .forEach(schema => (uniqueSchemas[schema] = true));
-
-        // Wait for all schemas to be created.
-        await Promise.all(
-          Object.keys(uniqueSchemas).map(schema => this.adapter.prepareSchema(schema))
-        );
-
-        await this.executeGraph();
-        resolve(this.result);
-      } catch (e) {
-        reject(e);
-      }
-    });
-
-    return this.executionTask;
+    this.executionTask = this.executeGraph();
+    return this.resultPromise();
   }
 
   public cancel() {
     this.cancelled = true;
     this.eEmitter.emit(CANCEL_EVENT);
   }
 
-  public resultPromise(): Promise<dataform.IExecutedGraph> {
-    return this.executionTask;
+  public async resultPromise(): Promise<dataform.IExecutedGraph> {
+    await this.executionTask;
+    return this.resultAsExecutedGraph();
   }
 
   private triggerChange() {
-    return Promise.all(this.changeListeners.map(listener => listener(this.result)));
+    const executedGraph = this.resultAsExecutedGraph();
+    return Promise.all(this.changeListeners.map(listener => listener(executedGraph)));
```

Comment:
feels like you are going to need another PR in this case to do the close-source switch. Would it not be better to expose resultAsExecutedGraph to user and change this API to it's final form here and now? Or you'll need some other way to add listeners for the actual RunResult object